jina/math/dimensionality_reduction.py

```diff
@@ -0,0 +1,56 @@
+import numpy as np
+
+
+class PCA:
+    """A class for PCA dimensionality reduction in Jina."""
+
+    def __init__(self, k: int, whitening: bool = True):
+        self.k = k
+        self.whitening = whitening
+
+    def fit(self, x_mat: np.ndarray):
+        """
+        Computes the projection matrix of the PCA algorithm and stores it to self.w
+
+        :param x_mat: Matrix of shape (n_observations, n_features)
+        """
+
+        x_mat = x_mat - x_mat.mean(axis=0)
+
+        # covariance matrix (n_features, n_features)
+        cov = np.cov(x_mat.T) / x_mat.shape[0]
+
+        # Compute eigenvalues eigenvectors of cov
+        e_values, e_vectors = np.linalg.eig(cov)
+
+        # Sort eigenvalues by magnitude (higher to lower)
+        idx = e_values.argsort()[::-1]
+        e_values = e_values[idx]
+
+        # Sort eigen vectors by eigenvalue magnitude (higher to lower)
+        e_vectors = e_vectors[:, idx]
+
+        # Projection matrix contains eigenvectors sorted
+        self.w = e_vectors
```

Comment:
Initialize these attributes at initialization time

Reply:
This is not known at initialization time, why set it up?

Reply:
set it to None perhaps, but initialize it.

Reply:
and in transform u can check if self w is not None.

Otherwise just using transform will not work

Reply:
done